cat/ptx-v7.5.cat

```diff
@@ -0,0 +1,91 @@
+PTX
+(* NVIDIA. Parallel Thread Execution ISA Version 7.5 *)
+
+// Base relations:
+// sr: same-scope
+// scta: same-cta
+// sync_barrier: synchronize by barriers with same logical barrier resource (barID)
+
+// Comparing to the ptx-v7.5 model written in Alloy (https://github.com/NVlabs/mixedproxy)
+// The ptx-v7.5 alloy model uses this Causality-axiom "irreflexive ((rf | co | fr)?; cause)".
+// The ptx-v6.0 alloy model uses "irreflexive ((rf | fr)?; cause)".
+// The ASPLOS'19 paper and the documentation refer to "rf | fr" (non-optional). 
+// We follow the later.
+
+(*******************)
+(*   Auxiliaries   *)
+(*******************)
+
+(* Events *)
+let strong-write = W & (RLX | REL)
+let strong-read = R & (RLX | ACQ)
+let strong-m = strong-write | strong-read
+let strong-operation = strong-m | F
+
+(* Proxy *)
+let same-proxy = GEN * GEN | SUR * SUR | TEX * TEX | CON * CON
+
+(* Relations *)
+let sr-fence = ([strong-operation]; (sr & same-proxy); [F]) | ([F]; (sr & same-proxy); [strong-operation])
+let po-fence = ([strong-operation]; (po & same-proxy); [F]) | ([F]; (po & same-proxy); [strong-operation])
+let sr-m = ([strong-m]; (sr & loc & same-proxy); [strong-m])
+let po-m = ([M]; (po & loc & same-proxy); [M])
+let morally-strong = sr-fence | po-fence | sr-m | po-m
+let fence-sc = [F & SC]; sr; [F & SC]
+
+// This is probably equivalent to 
+// let rec observation = (morally-strong & rf) | (observation; rmw; observation)
+// We opt to avoid the recursion
+let observation = (morally-strong & rf) | rmw
+
+let release-pattern = ([W & REL]; po-loc?; [strong-write]) | ([F & ACQ_REL]; po; [strong-write])
+
+let acquire-pattern = ([strong-read]; po-loc?; [R & ACQ]) | ([strong-read]; po; [F & ACQ_REL])
+
+let sync = morally-strong & (release-pattern; observation+; acquire-pattern)
+let cause-base = (po*; ((sync | sync_barrier | fence-sc); po*)+) | po
```

Comment:
1. Yes they are the same. The ptx-v6.0 was following the paper's style, it's updated to the same as here now for consistency.
2. The `cause` is meant to be the causality order relation, the name is following the [alloy model](https://github.com/NVlabs/mixedproxy/blob/main/alloy/ptx.als#L141) of NVIDIA's repository. 
3. Causality order captures how memory operations become visible across threads\proxies via synchronizing operations (release/acquire M and all kinds of fences or barriers). This relation is used in two axioms of the PTX memory model: in coherence axiom it restricts two writes in causality order must be in coherence order; in causality axiom it restricts the communication order(`rf`/`fr`) must be in causality order. Could you give an example how is it forming cycles?

Reply:
Regarding point 3: I assume the same-scope relation `sr` is an actual equivalence relation, i.e., it is symmetric.
Then a code as simple as 
```
P0
e1: sc-fence;
e2: someEvent;
e3: sc-fence;
```
causes `e2` to be related to itself via `e2 --po--> e3 --fence-sc-> e1 --po--> e2`.
Basically as soon as you place two SC-fences into some thread, all events inbetween are related to each other. Similar things happen with `sync_barrier` I believe.

Reply:
Btw. I was suggesting to change `po*` to `po?` in version 7.5, rather than the other way around.
The v6.0 paper also uses `po?`, so it matches better with the reference source.

Reply:
Thomas is right, we are missing these axioms from the the alloy models
- v6.0
```
//fence
fact wf_sc {
  //all s: Scope | total[sc, scoped[s] & inside[s] & FenceSC]
  all disj f1, f2 : FenceSC | f1->f2 in strong_r => (f1->f2 in sc or f2->f1 in sc)
  sc in typed[strong_r, FenceSC]
  acyclic[sc]
}
``` 
- v7.5
```
//sc
fact {
  all s: Scope | total[sc, scoped[s] & inside[s] & FenceSC]
}
```
In the later they go even further and require `sc` to be total. This is probably "wrong", i.e., not matching the documentation.

We need to make `sc` a base relation. It would be great if we could add test cases that show the current version if buggy.

Reply:
I don't think you latest commit solves the problems mentioned by Thomas. You just moved the same definition to the code rather than the cat file. I see two problems:
- you need to add to the cat file the fact that `sc` is acyclic (this should make some of your tests break; if not, we need more tests)
- the visitor that computes knowledge is adding both directions to the must set, meaning that both direction are always part of the relation (if the events execute). What you need to actually do is to add them to the may set, this + the acyclicity constraint will force the solver to add only one on the directions in many cases (making `sc` acyclic as the documentation states)

Reply:
Not only is it acyclic, it should also be total on morally strong sc fences:
```
The Fence-SC order is an acyclic partial order, determined at runtime, that relates every pair of morally strong fence.sc operations.
```
I think this requires a similar encoding to `co`, but possibly with different integer clock variables for different scopes/proxies (can `fence.sc` also be applied on different scope/proxy levels?).

Reply:
I don't think your semantics of `sync_fence` is right. You won't get away with modelling the properties of `sync_fence` inside the CAT file itself, because some of those properties are not expressible.
What you have right now is that `sync_fence` can take any value as long as it is acyclic.
But that also allows it to be simply empty in all executions, which is clearly wrong.

You will need to define a proper semantics in the `WmmEncoder` class. 
In particular, `sync_fence` has to be total over morally strong fences (which I think are fences in the same scope?), and this requires you to do something similar to how `co` is encoded.

Reply:
After adding a co-like encoder for sync_fence, there is a [test](https://github.com/tonghaining/Dat3M/blob/ptx-proxy/litmus/PTX/SB%2Bsc-gpu-multiFence.litmus) with multi SC getting different result on ptx-v6.0 and ptx-v7.5. Not sure if it's from incorrect encoder implementation or the model's difference?

Reply:
I tried on the alloy model with removing the `po` of `cause_base` and it got same result as ptx-v6.0. So I guess the difference comes from the models.

Reply:
So are you saying that the addition of `po` to `cause_base` in ptx-v7.5 might indeed give different semantics?
IIRC you were missing a concrete example that shows the semantics to be inconsistent with each other (according to what I got from the discussions on the NVIDIA github).
Fortunately, this should be rather easy to confirm. One of the models gives you a concrete execution which should or should not exist (not sure). You can take this exact execution and check it against the other model by hand.
It might take some effort to get the execution printed though (Dartagnan can do it, but it will only show the po, rf, and co, but not sync_fence or any derived relations).

Reply:
I just tried the current vresion of the code. v7.5 says the `forall` is guarantee and I think this is because the order between the stores in the same threads is guaranteed (due to `po` being part of `cause-base`). However I am not too sure v6.0 should guarantee this property (i.e. I could believe this tests show differences between the two models).

@tonghaining are you saying the following?
- in their alloy tool (for v7.5), the test passes, and
- in their alloy tool, but removing the `po` (we do this to mimic v6.0 which lacks tool support), the test also passes

Reply:
I don't understand how `forall` could fail to hold. To have a violating execution, you need both loads to read from the earlier store (or even the initial one). This gives you `fr`-edges from the loads to the later stores (those that store `2`). Since `sync_fence` is total, it will relate the two later fences in some direction (let's say from P0 to P1).
But then you get a `po;sync_fence;po` path from `st.weak x, 2` to `ld.weak r2, x`, i.e, you get a `cause` edge.
In the opposite direction you have the aforementioned `fr` edge, and so you get a `fr;cause` cycle which should be prohibited by `(* Axiom 6 Causality *) irreflexive ((rf | fr); cause) as axiom-Causality`

Reply:
The final part of your argument focus on the case where you read from the stores with value 2. What if you read from the initial ones? You would need a `cause` path from the init to the store with value 2 to forbid the execution, and I am not sure we have this.

Reply:
No, I say that you do NOT read from the store with value 2 and hence you get a `fr`-edge towards that store.
The fences guarantee that there is always a cause-path from one of the "value 2" stores to its corresponding load.
It shouldn't matter if there is a cause-path between the different stores.

Reply:
The `rf` edge would hold if not only you have the `rf` edge, but also the `co` one. I think this is one of the cases (because both stores are weak) where PTX does not guarantee you to have a `co` edge.

Reply:
You are right. If those two stores are not `co`-related, you indeed get different behaviour. But they have to be related according to the documentation:
```
Two overlapping write operations are related in coherence order if they are morally strong or if they are related in causality order
```
Both writes are overlapping (accessing the same address) and are morally strong (again, because they access the same address and are in program order). The SC-per-loc axiom prohibits `co` to oppose `po-loc`, so you get the forward coherences that are necessary to get the `fr;cause` cycle.

I think that our encoding of `co` is wrong and fails to relate morally strong accesses, and that is why we get different results here.

Reply:
Now I am confused. We are arguing that according to our reasoning, the `forall` should hold in both versions 6 and 7.5.

Dartagnan says it does not hold for v6, but we already identified what the problem is.

Are you saying their tool is also giving a wrong result? Have you added the missing axiom on top of removing `po` from `cause-base`?

Reply:
Okay it was from the missing `sc_per_location` axiom. I got the result from their main branch version (without the axiom). With adding it, both `v6` and `v7.5` will pass as tested.